### ISSUE TITLE
EpollEventLoop.wakeup takes mutch longer with new netty versions

### ISSUE THREAD

**Issue by SplotyCode** [netty/netty#10312]
While upgrading from netty 4.0.23 to 4.1.49 we experienced a much higher CPU consumption when sending data. We have 300-400 connections open and the server sends a few hundred small packets per second to every client.

After the upgrade, our main threads (that sends the packets using the method below) seem to 'hang' here: https://imgur.com/a/VMo9szT.

We send the packets like this:
```java
Channel channel = /* ... */;
Packet packet = /* ... */;
EventLoop eventLoop = channel.eventLoop();
if (eventLoop.inEventLoop()) {
  /* ... */
} else {
  eventLoop.execute(() -> {
    ChannelFuture future = channel.writeAndFlush(packet);
    future.addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE);
  });
}
```
We are using the default options except for TCP_NODELAY.
I saw that @njhill did multiple changes to the wait/wakeup logic could that be the cause of this?

**Comment by hyperxpro** [comment]
4.0.23 was released in August 2014 and 4.1.49 was released in April 2020. In 6 years, lots of things changed so higher CPU consumption is in 4.1.49 compared to 4.0.23 is pretty obvious.

Can't say if there is a problem in 4.1.49 or recent version until you post some benchmarks of both 4.0.23 and 4.1.49.

**Comment by njhill** [comment]
@SplotyCode a few questions:
- Do you know whether overall performance (latency etc) was better or worse despite the CPU increase?
- How about the rate of wakeups relative to rate of messages?
- Are you able to compare the performance with `NioEventLoop`?
- In your example code above, why do you schedule the writeAndFlush on the event loop? It can be called from outside the EL

One hunch is that it could be related to speed-ups on the event loop meaning it completes iterations faster and returns to wait prior to the next task coming in whereas before it may have stayed awake long enough to see it. You could experiment with adding a `Thread.yield()` after the write in the scheduled task (ignoring final bullet above)

**Comment by SplotyCode** [comment]
> so higher CPU consumption is in 4.1.49 compared to 4.0.23 is pretty obvious

I don't really get that. With that logic netty 3 would be the fastest version. I mean sure there are new features but I don't get why it would be so much higher. The profiler picture is from one hotspot and we have other hotspots all ending in wakeup. So independent of the versions it is unexpected to spend so long on wakeup right?

> @SplotyCode a few questions:

Now that I think about it, it was unbelievably rude to instantly 'blame' you when there are thousands of other possible commits and no evidence. I am really sorry.

> Do you know whether overall performance (latency etc) was better or worse despite the CPU increase?

Sadly we don't measure latency. We had more users complaining about high latency that we traced to this server but I don't know if it just affects a few users or all.

> How about the rate of wakeups relative to rate of messages?

I don't have permission to the CPU snapshot of production. I will ask and reply.

> Are you able to compare the performance with NioEventLoop?

We could not reproduce this issue other than in production. I will discuss if we can use NioEventLoop in production for a few hours.

> In your example code above, why do you schedule the writeAndFlush on the event loop? It can be called from outside the EL

I was wondering the same thing that is why I included it in the issue. The server software that we use is closed source (we do changes with decompiling locally and creating patch files). So generally I feel uncomfortable with changing internal stuff. (Maby this thread-safe behavior was changed in netty 4? The server was written in netty 3 and was ported badly in some places). There are a few edge cases were I believe they actually need this (i did not post full code) but this only affects < 0.01% of the packets. I made a PR that writes nearly all packets outside the event loop now. I will see if my colleges let me run that change in production.

> One hunch is that it could be related to speed-ups on the event loop meaning it completes iterations faster and returns to wait prior to the next task coming in whereas before it may have stayed awake long enough to see it. You could experiment with adding a Thread.yield() after the write in the scheduled task

Wow! That's a brilliant idea I will definitely try this after trying the other things mentioned.

I am so glad you answered @njhill. I was so done and now we have a bunch of new approaches.

**Comment by SplotyCode** [comment]
I now have information about wakeups relative to the rate of messages.
Since profilers use a lot of CPU we can not run them very long. In one hotspot EpollEventLoop#wakeup was called around 60k times. And 100% of these calls were delegated to Native#eventFdWrite. This is unexpected behavior right?

**Comment by njhill** [comment]
> This is unexpected behavior right?

Not really, it just means many tasks are submitted from outside the EL each of which must have completed prior to the next being submitted. This is consistent with my hypothesis above. Experimenting with the `Thread.yield()` suggestion might give more clues. How many other threads are submitting these tasks? An alternative would be to limit the rate that flush is called, per number of writes or number of microsecs (and probably per source thread). Calling write without flush won't wake up the EL.

It's up to you to decide latency/cpu tradeoff, the other extreme is to dedicate a core to the EL and use busy-wait.

BTW though it will reduce some overhead I don't expect your change to call writeAndFlush directly will make too much difference to the effects that you're observing since it still schedules a task on the EL itself.
